src/core/radio/radio.cpp

```diff
@@ -92,4 +95,130 @@ Error Radio::Transmit(Mac::TxFrame &aFrame)
 }
 #endif // OPENTHREAD_CONFIG_RADIO_LINK_IEEE_802_15_4_ENABLE
 
+#if OPENTHREAD_CONFIG_RADIO_TIME_STATS_ENABLE && (OPENTHREAD_FTD || OPENTHREAD_MTD)
+inline uint64_t UintSafeMinus(uint64_t aLhs, uint64_t aRhs)
+{
+    uint64_t result = 0;
+    if (aLhs > aRhs)
+    {
+        result = aLhs - aRhs;
+    }
+    return result;
+}
+
+RadioStatistics::RadioStatistics(void)
+    : mStatus(kDisabled)
+    , mSleepTime(0)
+    , mRxTime(0)
+    , mTxTime(0)
+    , mLastCalTime(TimerMicro::GetNow())
+{
+}
+
+void RadioStatistics::ChangeStatus(Status aStatus)
+{
+    UpdateTime();
+    mLastCalTime = TimerMicro::GetNow();
+    mStatus      = aStatus;
+}
+
+void RadioStatistics::HandleReceiveAt(uint32_t aDurationUs)
+{
+    // The actual rx time of ReceiveAt cannot be obtained from software level. This is a workaround.
+    if (mStatus == kSleep)
+    {
+        mRxTime += aDurationUs;
+    }
+}
+
+void RadioStatistics::TxDone(otError aError, uint16_t aPsduLength)
+{
+    if (aError == kErrorNone || aError == kErrorNoAck)
+    {
+        uint32_t txTimeUs = (aPsduLength + Mac::Frame::kPhyHeaderSize) * Radio::kSymbolsPerOctet * Radio::kSymbolTime;
+        uint32_t rxAckTimeUs = (Mac::Frame::kImmAckLength + Mac::Frame::kPhyHeaderSize) * Radio::kPhyUsPerByte;
```

Comment:
I think we are assuming here that radio platform will do a single frame tx and won't be doing the retries?

If radio does retries (which RCP and most radio platform do) then this won't be correct.

Reply:
Yeah, for RCP (posix) this is not correct. My plan is to handle the posix case in a different way later. Considering this feature is mainly for FTD, MTD (we want to know the percentage on power constrained devices), maybe let's put some comment in the code to clarify this isn't accurate on posix.

What other platforms will have this issue? As I understand, expect for posix+rcp, the radio callback is called each time for a single frame tx.

Reply:
I think when radio platform does handle retries it won't be correct (under `OT_RADIO_CAPS_TRANSMIT_RETRIES`). In such a case we get the `TxDone` once potentially after some number of reties (e.g. can have two tx attemps with no-ack and third attempt being successful).

I think most radio platform do provide `OT_RADIO_CAPS_TRANSMIT_RETRIES`?

Reply:
Oh I see! 

I find something like `nrf_802154_tx_started` in nRF's radio platform and that should be called for each single frame tx. But currently we don't have a callback to call from this ISR and it seems we shouldn't do such thing (may cause racing issue).

Currently I cannot think of any solutions for this issue. Maybe let's assume there will be only a single frame tx and make it clear in the doc. Just like `TxRetry` in MacCounters, it's not accurate when `OT_RADIO_CAPS_TRANSMIT_RETRIES` is enabled. For more accurate counting, I think we can further work on #8820 and design some platform APIs. Thoughts?

Reply:
Would be great if we can get a callback on each frame tx from radio platform? (but it requires new radio plat APIs and changes, not easy to add likely). We have something like this in `SubMac` callbacks -> `RecordFrameTransmitStatus()` and `RecordCcaStatus()`.

Reply:
I agree it would be good to have a callback on each single frame tx. In terms of this PR, I think it may not be worthwhile because adding new radio plat APIs. The intention is to implement this feature without minimal support from the radio platform. Adding a new radio plat API requires all platform to support this which seems against the original purpose.

Reply:
I fully agree, These are ideas for future (not in this PR).

Reply:
Perhaps its already here somewhere and I missed it but is there a way to prevent this code from being included in RCP builds since it is not accurate for RCPs?